File: src/RegexLearningGuide/Helpers/DemoHelper.cs
using System.Text.RegularExpressions;

namespace RegexLearningGuide.Helpers;

/// <summary>
/// Utility methods to display regex results in a clean, educational format.
/// </summary>
public static class DemoHelper
{
    public static void ShowExample(string title, string pattern, string input,
        RegexOptions options = RegexOptions.None)
    {
        Console.WriteLine($"  ▸ {title}");
        Console.WriteLine($"    Pattern : {pattern}");
        Console.WriteLine($"    Input   : \"{input}\"");

        var regex = new Regex(pattern, options);
        var matches = regex.Matches(input);

        if (matches.Count == 0)
        {
            Console.WriteLine("    Result  : No match found");
        }
        else
        {
            foreach (Match match in matches)
            {
                Console.WriteLine($"    Match   : \"{match.Value}\" at index {match.Index}");

                // Show groups if there are capturing groups (skip group 0 = full match)
                if (match.Groups.Count > 1)
                {
                    for (int i = 1; i < match.Groups.Count; i++)
                    {
                        var group = match.Groups[i];
                        string name = group.Name == i.ToString() ? $"Group {i}" : $"Group '{group.Name}'";
                        Console.WriteLine($"              {name}: \"{group.Value}\"");
                    }
                }
            }
        }

        Console.WriteLine();
    }

    public static void ShowIsMatch(string title, string pattern, string input,
        RegexOptions options = RegexOptions.None)
    {
        Console.WriteLine($"  ▸ {title}");
        Console.WriteLine($"    Pattern : {pattern}");
        Console.WriteLine($"    Input   : \"{input}\"");
        Console.WriteLine($"    IsMatch : {Regex.IsMatch(input, pattern, options)}");
        Console.WriteLine();
    }

    public static void ShowReplace(string title, string pattern, string input,
        string replacement, RegexOptions options = RegexOptions.None)
    {
        Console.WriteLine($"  ▸ {title}");
        Console.WriteLine($"    Pattern     : {pattern}");
        Console.WriteLine($"    Input       : \"{input}\"");
        Console.WriteLine($"    Replacement : \"{replacement}\"");
        Console.WriteLine($"    Result      : \"{Regex.Replace(input, pattern, replacement, options)}\"");
        Console.WriteLine();
    }

    public static void ShowSplit(string title, string pattern, string input,
        RegexOptions options = RegexOptions.None)
    {
        Console.WriteLine($"  ▸ {title}");
        Console.WriteLine($"    Pattern : {pattern}");
        Console.WriteLine($"    Input   : \"{input}\"");
        var parts = Regex.Split(input, pattern, options);
        Console.WriteLine($"    Parts   : [{string.Join(", ", parts.Select(p => $"\"{p}\""))}]");
        Console.WriteLine();
    }

    public static void Section(string title)
    {
        Console.WriteLine($"  ── {title} ──");
        Console.WriteLine();
    }

    public static void Note(string text)
    {
        Console.WriteLine($"  💡 {text}");
        Console.WriteLine();
    }
}

File: ts-regex-guide/src/helpers/demo-helper.ts
// ============================================================================
// Demo Helper — Utility functions for displaying regex results clearly
// ============================================================================

export function section(title: string): void {
  console.log(`  ── ${title} ──`);
  console.log();
}

export function note(text: string): void {
  console.log(`  💡 ${text}`);
  console.log();
}

export function showExample(
  title: string,
  pattern: RegExp,
  input: string
): void {
  console.log(`  ▸ ${title}`);
  console.log(`    Pattern : ${pattern}`);
  console.log(`    Input   : "${input}"`);

  const matches = [...input.matchAll(new RegExp(pattern.source, pattern.flags.includes("g") ? pattern.flags : pattern.flags + "g"))];

  if (matches.length === 0) {
    console.log(`    Result  : No match found`);
  } else {
    for (const match of matches) {
      console.log(`    Match   : "${match[0]}" at index ${match.index}`);

      // Show numbered groups
      for (let i = 1; i < match.length; i++) {
        if (match[i] !== undefined) {
          console.log(`              Group ${i}: "${match[i]}"`);
        }
      }

      // Show named groups
      if (match.groups) {
        for (const [name, value] of Object.entries(match.groups)) {
          if (value !== undefined) {
            console.log(`              Group '${name}': "${value}"`);
          }
        }
      }
    }
  }

  console.log();
}

export function showTest(
  title: string,
  pattern: RegExp,
  input: string
): void {
  console.log(`  ▸ ${title}`);
  console.log(`    Pattern : ${pattern}`);
  console.log(`    Input   : "${input}"`);
  console.log(`    Test    : ${pattern.test(input)}`);
  console.log();
}

export function showReplace(
  title: string,
  pattern: RegExp,
  input: string,
  replacement: string | ((substring: string, ...args: any[]) => string)
): void {
  console.log(`  ▸ ${title}`);
  console.log(`    Pattern     : ${pattern}`);
  console.log(`    Input       : "${input}"`);
  if (typeof replacement === "string") {
    console.log(`    Replacement : "${replacement}"`);
    console.log(`    Result      : "${input.replace(pattern, replacement)}"`);
  } else {
    console.log(`    Replacement : <function>`);
    console.log(`    Result      : "${input.replace(pattern, replacement)}"`);
  }
  console.log();
}

export function showSplit(
  title: string,
  pattern: RegExp,
  input: string
): void {
  console.log(`  ▸ ${title}`);
  console.log(`    Pattern : ${pattern}`);
  console.log(`    Input   : "${input}"`);
  const parts = input.split(pattern);
  console.log(`    Parts   : [${parts.map((p) => `"${p}"`).join(", ")}]`);
  console.log();
}
